scheduler/src/cook/impersonation.clj

```diff
@@ -0,0 +1,84 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns cook.impersonation
+  "Support for services impersonating a user, performing Cook actions the user's behalf."
+  (:require [clojure.string]
+            [clojure.tools.logging :as log]
+            [ring.util.response :refer (header status response)])
+  (:import (clojure.lang Keyword)))
+
+(defn- unrestricted-access
+  "Denotes that an impersonator may perform any operation on a given object type."
+  [_]
+  true)
+
+(defn- getter-access
+  "Denotes that an impersonator may only perform gets on a given object type."
+  [verb]
+  (= :get verb))
+
+(def object-type->verb->impersonatable?
+  "Specification of operations that authorized-impersonators may perform on behalf of another user."
+  {:job unrestricted-access
+   :quota getter-access
+   :share getter-access
+   :usage getter-access})
+
+(defn impersonation-authorized-wrapper
+  "This function is meant to wrap the authorization/is-authorized? function.
+
+  The resulting function determines whether an authorized impersonator
+  can perform the given operation on the given object, on behalf of the given user.
+
+  Returns true if allowed, else false."
+  [is-authorized-fn settings]
+  (fn is-impersonatable?
+    [^String user
+     ^Keyword verb
+     ^String impersonator
+     {:keys [owner item] :as object}]
+    (log/debug "[is-impersonatable?] Checking whether user" impersonator
+               "may impersonate user" user
+               "performing" verb "on" (str object) "...")
+    (log/debug "[is-impersonatable?] Settings are:" settings)
+    (and (or (nil? impersonator)
+             ((object-type->verb->impersonatable? item) verb))
+         (is-authorized-fn settings user verb impersonator object))))
+
+(defn create-impersonation-middleware
+  "Provides user-impersonation middleware.
+  Uses the authorized-impersonators set argument to verify impersonation authorization."
+  [authorized-impersonators]
+  (fn impersonation-middleware [h]
+    (if-let [impersonators-set (some-> authorized-impersonators seq set)]
+      (fn impersonation-wrapper [{user :authorization/user :as req}]
+        (if-let [impersonated-principal (get-in req [:headers "x-cook-impersonate"])]
+          (let [[impersonated-user] (clojure.string/split impersonated-principal #"[/@]" 2)]
```

Comment:
@sradack suggests that we should have all the principal → username conversions calling the same function. Just truncating the principal is a nasty hack, and we may want to do something more robust later, in which case we don't want to be hunting all over our code to find these.